typing/typemod.ml

```diff
@@ -707,6 +708,77 @@ let merge_constraint initial_env loc sg lid constr =
   with Includemod.Error explanation ->
     raise(Error(loc, initial_env, With_mismatch(lid.txt, explanation)))
 
+(* A simplified version of [merge_constraint] for the case of packages. Package
+   constraints are much simpler - they must be "with type" constraints for types
+   with no parameters and can only update abstract types. *)
+let merge_package_constraint initial_env loc sg lid cty =
+  let rec patch_item namelist outer_sig_env sg_for_env ~ghosts item =
+    let return replace_by =
+      Some ((), {Signature_group.ghosts; replace_by})
+    in
+    match item, namelist with
+    | Sig_type(id, sig_decl, rs, priv) , [s]
+      when Ident.name id = s ->
+        begin match sig_decl.type_manifest with
+        | None -> ()
+        | Some ty ->
+          raise (Error(loc, outer_sig_env, With_package_manifest (lid.txt, ty)))
+        end;
+        let new_sig_decl =
+          Typedecl.transl_package_constraint ~loc outer_sig_env cty.ctyp_type
+        in
+        check_type_decl outer_sig_env sg_for_env loc id None
+          new_sig_decl sig_decl;
+        let new_sig_decl = { new_sig_decl with type_manifest = None } in
+        return (Some(Sig_type(id, new_sig_decl, rs, priv)))
+    | Sig_module(id, _, md, rs, priv) as item, s :: namelist
+      when Ident.name id = s ->
+        let sig_env = Env.add_signature sg_for_env outer_sig_env in
+        let sg = extract_sig sig_env loc md.md_type in
+        let ((), newsg) = merge_signature sig_env sg namelist in
+        let item =
+          match md.md_type with
+            Mty_alias _ ->
+              (* A module alias cannot be refined, so keep it
+                 and just check that the constraint is correct *)
+              item
+          | _ ->
+              let newmd = {md with md_type = Mty_signature newsg} in
+              Sig_module(id, Mp_present, newmd, rs, priv)
+        in
+        return (Some item)
+    | _ -> None
+  and merge_signature env sg namelist =
+    match
+      Signature_group.replace_in_place (patch_item namelist env sg) sg
+    with
+    | Some (x,sg) -> x, sg
+    | None -> raise(Error(loc, env, With_no_component lid.txt))
+  in
+  try
+    let names = Longident.flatten lid.txt in
+    let (tcstr, sg) = merge_signature initial_env sg names in
+    check_well_formed_module initial_env loc "this instantiated signature"
+      (Mty_signature sg);
+    (tcstr, sg)
+  with Includemod.Error explanation ->
+    raise(Error(loc, initial_env, With_mismatch(lid.txt, explanation)))
+
+let check_package_with_type_constraints loc env mty constraints =
+  let sg = extract_sig env loc mty in
+  let sg =
+    List.fold_left
+      (fun sg (lid, cty) ->
+         snd (merge_package_constraint env loc sg lid cty))
```

Comment:
Why not just call `merge_constraint` here? That's definitely overkill, but it seems better than duplicating code. I don't see what `merge_package_constraint` does that's better than `merge_constraint`.

Reply:
I don't think this works - trying to use `merge_constraint` for both things is the source of the problems fixed in this PR.

The key difference is that `merge_package_constraint` takes the constraint as an already-translated bit of typedtree, while `merge_constraint` takes it as a not-yet-translated parsetree type decl.  That type decl is subsequently translated in the wrong type variable environment to support package constriants, which is why the old version used the fake malformed bit of parsetree described in the PR description above.  I tried to explain there why obvious small tweaks to that approach don't work - maybe that makes more sense after this answer, but if not let me know and I'll try again.

Reply:
I had missed the difference between the parsetree and the typedtree here -- an important distinction.

But I'm still not convinced we have to repeat all this code. You encountered two problems, but I think both can be addressed without duplicating the code:

1. We must make sure that the type in the original signature is abstract. (Why? I'm not sure. But let's not try to fix that today.)
2. We must allow variables to be used from the environment.

It seems relatively easy to accommodate both (1) and (2) with an extra parameter passed to `merge_constraint`, saying whether we're in a package type. For (1), we just do an extra check. For (2), we skip the `TyVarEnv.reset()` in `transl_with_constraint`. If that runs into trouble, then maybe `merge_constraint` decides between calling `transl_with_constraint` and the new `transl_package_constraint` (which I'm not bothered by, because it's so short).

It's possible there will be more obstacles on this path. I'm just concerned by duplicating so much code here. Also happy to chat about this live this afternoon, say.

Reply:
I fairly strongly disagree.  Maybe that would work, maybe it would not.  (In practice I did try deleting the `TyVarEnv.reset` call you mention and that's not enough to get the correct tyvarenv.  I did not deeply investigate why, but this contributes to my view that the old approach is very confusing and is best to ditch.)

Even if we could easily additionally parameterize `transl_with_constraint`, I think approach I've proposed here is much better!  I believe this PR considerably clarifies how package with constraints work. The old approach, of constructing a fake bit of parse tree (Bad!) and having a bunch of subtle special cases, was pretty confusing.  It took me quite a while to work out what is allowed and what is not allowed for package constriants.  In the new version, that's it's own function and you can just read it.

Reply:
If the aim is to merge the `merge_constraint` function, an option that could probably work would be to add a `With_type_package` variant to the `with_info` type which takes a typed type declaration rather than a parsetree fragment. Then `merge_package_constraint` could be:
```ocaml
let merge_package_constraint initial_env loc sg lid cty =
  let _, s = merge_constraint initial_env loc sg lid (With_package_type cty) in
  s
```
(A dirty implentation is visible at https://github.com/ocaml/ocaml/commit/7c0ebd30b6de17f412fe4f0486f88d8d4a1d0fe8)

However, I have mixed feeling about this idea: yes, the complex code of `merge_constraint` is shared, but this code is complex because it is mixing multiple functions with a small common spine. Thus I am not completely sure if this version is simpler than the version which is extracting the simple package type case from the chimeric aggregated function.

Reply:
> If the aim is to merge the `merge_constraint` function, an option that could probably work would be to add a `With_type_package` variant to the `with_info` type which takes a typed type declaration rather than a parsetree fragment.

I like this idea!  It nicely isolates the behavior of package with type constraints so it's less spread out in special cases of `merge_constraint` and `transl_with_constraint`, and eliminates the need to construct fake parse tree.  I will give it a try.

Reply:
I have pushed a commit implementing @Octachron's idea.

Compared with his draft, I made a few changes:
1) I made a separate case in `merge_constraint` for `With_type_package`, rather than combining it into the `With_type`/`With_typesubst` case.  I think this is clearer, even if it results in a couple duplicated lines.
2) I changed `merge_constraint` to only conditionally return the typed tree `with_constraint` (by returning an option).  This constraint was unused and slightly nonsense in the `With_type_package` constraint case, so this seems clearer.
3) I kept the requirement that package with constraints only work on type declarations without manifests.  A nice thing about his version is that we could potentially allow updating manifests with compatible definitions, but I believe this will require more thought around row variables (or at least more work to get good error messages in those cases), so I propose to leave it as future work.  I have added a few more tests of this type.  (Note this PR preserves today's behavior, it just gives it a clearer error message.)

I agree with @Octachron that there are trade offs here - the old version was maybe a little clearer by having two separate functions rather than combining them into one just to avoid duplicating the "common spine".  But I think this approach is probably the best compromise, eliminating the need to build fake parsetree without duplicating too much code.